sigs/jvm/rfcs/20191023-tensorflow-keras.md

```diff
@@ -0,0 +1,342 @@
+
+Combining the wide support of the JVM ecosystem and development environment with a 
+practical interface to TensorFlow Java will allow users to easily integrate deep learning
+into their existing development and production workflows.
+
+## Design Proposal
+
+In building this framework, the goal is to closely match the behaviors and workflows
+Keras supports in its Python API, with a well designed Java API that can be easily
+integrated into existing projects.
+
+### Overview
+
+Keras is built around a few core abstractions which comprise much of what is needed to build
+deep learning models. These include layers, models, optimizers, activations, losses, metrics,
+regularizers, and initializers.
+
+To make our code clearer and more manageable, we distinguish explicitly between two steps:
+Graph building (even in eager mode, graph "definition"), and Graph execution.
+
+
+
+### Layers
+Layers represent arbitrary graph computations on tensors. They are composable; meant to be "stacked"
+on each other to build neural networks. Thus they must encode *differentiable* functions (i.e. built
+using TF Operations).
+
+Layers are defined by (1) a set of weights, and (2) a differentiable "transform" function which uses
+these weights to transform input tensor(s). For example, a `Dense` layer contains a kernel
+weight matrix *w_k* and a bias weight matrix *w_b*, and transforms an input *x* using the
+function *x* -> *f*(*w_k* x + *w_b*) for a user-specified activaition *f*.
+
+#### Defining Layers
+To define a new type of layer, users can override 3 methods from the base `Layer<T>` class.
+These are: 
+
+1. `Layer.build(Ops tf, Shape inputShape)`
+
+    The `build` method is intended for creating tensorflow variables which
+    will be required to compute the output of a layer. For example, for a
+    Dense layer implementation, `build` is used to create the kernel (*w_k*) and
+    bias (*w_b*) weight matrices.
+
+    To add weight variables to a layer, subclasses overriding the `build` method must call
+    the protected method `Layer.addWeight`; an internal wrapper for `tf.variable`.
+    See the Dense layer implementation for an example of this usage.
+
+    There are two build methods; one that should be overridden in any `Layer` subclass,
+    and a public method which wraps the overridden method, to be called by external
+    users of any Layer class. 
+    ```java
+    // Override this in Layer subclasses. To access dtype
+    // within this method, call this.getDtype()
+    protected abstract void build(Ops tf, Shape inputShape)
+
+    // Call this method when using a created Layer object externally.
+    // Requires the user to provide a dtype, as some tensorflow 
+    // operations require the explicit specification of a dtype
+    // (e.g. tf.variable(Shape shape, Class<T> dtype) and tf.placeholder(Class<T> dtype))
+    public final void build(Ops tf, Shape inputShape, Class<T> dtype)
+    ```
+2. `Layer.call(Ops tf, Operand<T>... inputs)`
```

Comment:
I'm wondering if there is a gain to pass the `Ops` instance to every call of the `Layer` interface. Since `Layer` already seems to be a stateful API, can't we store the `Ops` that was use for creating it and reuse on subsequent call?

Also, maybe we could avoid to pass it at all by letting the Keras API use the default eager environment unless specified otherwise in some initializer, something like `Keras.initEnvironment(Ops tf)`? I'm not sure there is a gain for the user to keep direct access to the `Ops` instance.

Reply:
I have experimented with a few options for managing this. The available options seem to be

1. Leave the management of `ExecutionEnvironment` and `Ops` instances completely up to the user; on the API side, this means accepting `tf` or `ExecutionEnvironment` as a function argument anywhere it is needed.

Reply:
Good points here, Karl, thanks! I actually experimented with a couple different approaches like you outlined.

 My feeling recently is that management of `ExecutionEnvironment` and `Ops` might actually be something to design / solve at the `TensorFlow` level, rather just at the `Keras` level, since this is a usage pattern every user of TF-Java will have to manage, even if not using a sub-framework like Keras.

The options available, as far as I can tell, are to

(1) Leave the management of these objects completely to the user. On the framework side, this means exposing `ExecutionEnvironment` and `Ops` instances as arguments to user-callable functions when needed.

This is the approach I took so far. I definitely agree with your point regarding `Layer`.  Passing in `Ops` only once to `Layer.build`, and storing a reference for re-use during `Layer.call` is a good idea, since you're never supposed to call `build` and `call` with different `ExecutionEnvironment` objects.

(2) Manage a global `ExecutionEnvironment` statically (maybe in the `TensorFlow` class). This way, even within internal framework methods, you can always get the current `ExecutionEnvironment` or `Ops` instance with a call like `TensorFlow.getCurrentEnvironment()` or `TensorFlow.ops()`. 

A strength, perhaps, of TF-Java right now is that you can work with multiple `ExecutionEnvironment` objects at the same time; this solution wouldn't really allow this, which is why I went with the first solution. admittedly, there's probably not a ton of use for this, but given that it exists, we can enable such use?

(3) Simulate the a Python session context manager.  A Pythonic way of handling this would be to create an internal "stack" of `ExecutionEnvironment` objects which are referenced by a context manager, so you can have something like:

```
with EagerSession() as env1:
   ... {TensorFlow code using env1}
   with EagerSession() as env2:
       ... {TensorFlow code using env2}
   ... {TensorFlow code using env1}
```

In Java, there are a couple ways I could see of emulating these semantics:

1. Using a try-with-resources and wrapper methods for generating `ExecutionEnvironment Objects, you could do something like the following:

```
class TensorFlow {
   private static Stack<ExecutionEnvironment> environments = new Stack();
   
   public static EagerSession getEagerSession() {
      environments.push(new EagerSession());
      return environments.top();
   }

   public static Graph getGraph() {
      environments.push(new Graph());
      return environments.top();
   }

   public static Ops ops() {
      return Ops.create(environments.top());
   }
}

// On EagerSession.close, call TensorFlow.environments.pop().close()
try (EagerSession eagerSession1 = TensorFlow.getEagerSession()) {
   try (EagerSession eagerSession 2= TensorFlow.getEagerSession()) {
      Ops tf = TensorFlow.ops();
   } // Pop & Close eagerSession2, 
} // Pop & Close eagerSession1
```

2. Or, we could use a wrapped function called something like `withEnvironment`, which takes an `ExecutionEnvironment`, and a `Lambda` accepting an `Ops` instance, which runs the lambda with this context; composing these functions could approximate the Python semantics
```
withEnvironment(new EagerSession(), tf -> {
   withEnvironment(new EagerSession(), tf -> { ... using inner env })
   ... using outer env
})
```

Should we add something like this to TensorFlow? Would love to know your thoughts.